File: backend/app/core/text_utils.py
"""Text cleaning and normalization utilities for HIGEN-ABSA."""

from __future__ import annotations

import re


def clean_text(text: str) -> str:
    """Clean raw review text for model inference.

    Strips leading 'Review N:' prefixes and collapses whitespace.
    """
    text = re.sub(r"^\s*Review\s+\d+\s*:\s*", "", text, flags=re.IGNORECASE)
    text = re.sub(r"\s+", " ", text)
    return text.strip()


def trim_span(text: str, start: int, end: int) -> tuple[int, int, str]:
    """Trim leading/trailing punctuation and whitespace from a text span."""
    trim_chars = " \t\r\n,.;:!?()[]{}\"'"
    while start < end and text[start] in trim_chars:
        start += 1
    while end > start and text[end - 1] in trim_chars:
        end -= 1
    return start, end, text[start:end]


CLAUSE_BOUNDARY_PATTERN = re.compile(
    r"[,;.!?]+|\b(?:nhưng|nhung|nhg|tuy\s+nhiên|tuy\s+nhien|mà|ma|còn|con)\b",
    re.IGNORECASE,
)


def clause_span_for_match(text: str, start: int, end: int) -> tuple[int, int, str]:
    """Expand a match span to clause boundaries for better evidence extraction."""
    left = 0
    right = len(text)
    for boundary in CLAUSE_BOUNDARY_PATTERN.finditer(text):
        if boundary.end() <= start:
            left = boundary.end()
            continue
        if boundary.start() >= end:
            right = boundary.start()
            break
    left, right, evidence = trim_span(text, left, right)
    if evidence:
        return left, right, evidence
    return trim_span(text, start, end)

File: backend-dotnet/HigenAbsa.Api/Core/TextUtils.cs
// Core/TextUtils.cs - Port of text_utils.py
using System.Text.RegularExpressions;

namespace HigenAbsa.Api.Core;

public static class TextUtils
{
    private static readonly Regex ReviewPrefixPattern =
        new(@"^\s*Review\s+\d+\s*:\s*", RegexOptions.IgnoreCase | RegexOptions.Compiled);
    private static readonly Regex WhitespacePattern =
        new(@"\s+", RegexOptions.Compiled);

    /// <summary>Clean raw review text: strip Review N: prefix and collapse whitespace.</summary>
    public static string CleanText(string text)
    {
        text = ReviewPrefixPattern.Replace(text, "");
        text = WhitespacePattern.Replace(text, " ");
        return text.Trim();
    }

    private static readonly Regex ClauseBoundaryPattern = new(
        @"[,;.!?]+|\b(?:nhưng|nhung|nhg|tuy\s+nhiên|tuy\s+nhien|mà|ma|còn|con)\b",
        RegexOptions.IgnoreCase | RegexOptions.Compiled);

    private const string TrimChars = " \t\r\n,.;:!?()[]{}\"'";

    /// <summary>Trim leading/trailing punctuation and whitespace from a text span.</summary>
    public static (int Start, int End, string Evidence) TrimSpan(string text, int start, int end)
    {
        while (start < end && TrimChars.Contains(text[start])) start++;
        while (end > start && TrimChars.Contains(text[end - 1])) end--;
        return (start, end, text[start..end]);
    }

    /// <summary>Expand a match span to clause boundaries for evidence extraction.</summary>
    public static (int Start, int End, string Evidence) ClauseSpanForMatch(string text, int matchStart, int matchEnd)
    {
        int left = 0;
        int right = text.Length;

        foreach (Match boundary in ClauseBoundaryPattern.Matches(text))
        {
            if (boundary.Index + boundary.Length <= matchStart)
            {
                left = boundary.Index + boundary.Length;
                continue;
            }
            if (boundary.Index >= matchEnd)
            {
                right = boundary.Index;
                break;
            }
        }

        var (ts, te, evidence) = TrimSpan(text, left, right);
        if (!string.IsNullOrEmpty(evidence))
            return (ts, te, evidence);

        return TrimSpan(text, matchStart, matchEnd);
    }
}
